### PR TITLE
feat(gpt-oss): Add CPU offload optimizer, differential LR/WD, and more

### PR DESCRIPTION
**Pull request by eous**
This PR introduces several features for memory-efficient fine-tuning of large models:

 ## CPU Offload Optimizer
- New `CPUOffloadOptimizersContainer` using DeepSpeed's CPUAdam
- Keeps optimizer states (momentum, variance) on CPU to reduce GPU memory
- Supports DTensor/FSDP2 with proper gradient redistribution
- Includes `cleanup()` method for releasing pinned memory resources
- Shape validation for gradient buffers with automatic reallocation

 ## Differential Learning Rates & Weight Decay
- New `LRMultipliers` and `WeightDecayMultipliers` config dataclasses
- Per-parameter-group settings: embeddings, output, attention, experts, routers, norms, mhc, bias
- `OptimizersContainerWithParamGroups` for standard (non-CPU-offload) differential training
- Pre-compiled regex patterns for efficient parameter classification

 ## HuggingFace Checkpoint Loading
- Direct safetensors loading bypassing dcp.load() pickle issues with DTensors
- Rank 0 broadcast pattern for distributed loading efficiency
- Proper handling for both distributed and non-distributed modes
- Uses configured `export_dtype` instead of hardcoded bfloat16
- Filter quantization tensor keys when converting unquantized checkpoints

 ## Loss Function
- Added `ignore_index=-100` to cross_entropy_loss for padding token support
- Exported `CROSS_ENTROPY_IGNORE_INDEX` constant

 ## Code Quality Improvements
- Extracted `_load_hf_state_dict_to_model()` helper to reduce duplication
- Added `Counter` for efficient dtype counting
- Named constants for magic numbers
- Documented private `_local_tensor` API usage with noqa comments

 ## Reference Configuration
- Added `persona_zeta_49k.toml` as known good config for GPT-OSS 20B fine-tuning
- Demonstrates CPU offload, differential LR/WD, 49K context, sample packing
- Validated on 2x nvidia rtx pro 6000 blackwell gpu's with FSDP sharding

### CODE REVIEW

File: torchtitan/models/gpt_oss/train_configs/20b_finetune_reference.toml [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+#
+# This is a reference configuration for fine-tuning the GPT-OSS 20B model.
+# It demonstrates the recommended settings for:
+#   - CPU offload optimizer (for memory efficiency on limited GPU memory)
+#   - Differential learning rates (per parameter group)
+#   - Differential weight decay (standard practice for transformers)
+#   - Long context training (49K sequence length)
+#   - Sample packing for efficiency
+#
+# This config was validated on 2x GPUs with FSDP sharding.
```
[the remark this answers is not in the document]
Comment (eous) [reply torchtitan/models/gpt_oss/train_configs/20b_finetune_reference.toml]: The NVIDIA RTX PRO 6000 Blackwell is a real product - it's NVIDIA's professional workstation GPU based on the Blackwell architecture (announced late 2024). The reviewer appears to be confusing product lines:
  - GeForce RTX = Consumer gaming
  - RTX PRO (formerly Quadro) = Professional workstation  
  - H100/B100 = Datacenter

  The RTX PRO 6000 Blackwell sits in the professional workstation tier, not datacenter. This config was validated on 2x RTX PRO 6000 Blackwell GPUs.

File: torchtitan/components/optimizer.py [hunk 2]
```diff
@@ -131,6 +324,88 @@ def _post_init(
         Optimizer.__init__(self, all_params, optimizer_kwargs)
 
 
+class OptimizersContainerWithParamGroups(OptimizersContainer):
```
[the remark this answers is not in the document]
Comment (eous) [reply torchtitan/components/optimizer.py:357]: The base class __init__ is intentionally not called because the subclasses have different optimizer creation logic:
  - OptimizersContainer creates one optimizer per model_part
  - OptimizersContainerWithParamGroups creates a single optimizer with param groups
  - CPUOffloadOptimizersContainer creates CPU-based optimizers with shadow params

  However, all subclasses:
  1. Manually set required attributes (self.optimizers, self.model_parts)
  2. Call self._post_init() which invokes Optimizer.__init__(self, all_params, optimizer_kwargs)

  This ensures proper initialization of the Optimizer base class functionality (hooks, state tracking) while allowing different optimizer creation strategies. See lines 375, 448, and 602.

File: torchtitan/components/optimizer.py [hunk 3]
```diff
@@ -181,6 +456,322 @@ def zero_grad(self) -> None:
         pass
 
 
+class CPUOffloadOptimizersContainer(OptimizersContainer):
```
[the remark this answers is not in the document]
Comment (eous) [reply torchtitan/components/optimizer.py:417]: The base class __init__ is intentionally not called because the subclasses have different optimizer creation logic:
  - OptimizersContainer creates one optimizer per model_part
  - OptimizersContainerWithParamGroups creates a single optimizer with param groups
  - CPUOffloadOptimizersContainer creates CPU-based optimizers with shadow params

  However, all subclasses:
  1. Manually set required attributes (self.optimizers, self.model_parts)
  2. Call self._post_init() which invokes Optimizer.__init__(self, all_params, optimizer_kwargs)

  This ensures proper initialization of the Optimizer base class functionality (hooks, state tracking) while allowing different optimizer creation strategies. See lines 375, 448, and 602.

File: torchtitan/components/optimizer.py [hunk 4]
```diff
@@ -17,19 +17,212 @@
...
+# Pre-compiled regex patterns for parameter group classification
+# IMPORTANT: Pattern order matters!
+# - 'bias' must be FIRST to catch all bias parameters before other patterns
+# - 'mhc' must come before 'norms' since mhc modules contain norm submodules
+# Note: norm.weight is the final RMSNorm before output, grouped with output
+# Patterns handle optional _checkpoint_wrapped_module wrapper from activation checkpointing
+import re
+_PARAM_GROUP_PATTERNS = {
+    # Bias pattern FIRST - catches all bias parameters
+    'bias': re.compile(r'\.bias$'),
```
[the remark this answers is not in the document]
Comment (eous) [reply torchtitan/components/optimizer.py]: This is intentional. The pattern matches all bias parameters because standard practice in transformer training is to disable weight decay on all bias terms, not just specific ones. This is well-established (see AdamW paper, HuggingFace Trainer defaults, PyTorch examples). The pattern is placed first in the ordered dict specifically to catch biases before any other pattern matches.

File: torchtitan/models/moe/moe.py [hunk 5]
```diff
@@ -303,48 +317,59 @@ def forward(
...
-        # By default, sigmoid or softmax is performed in float32 to avoid loss explosion
-        if self.score_func == "sigmoid":
-            scores = torch.sigmoid(scores.to(torch.float32))
-        elif self.score_func == "softmax":
-            scores = F.softmax(scores.to(torch.float32), dim=1)
+        # Get raw router logits shape (bs*slen, num_experts)
+        router_logits = self.gate(x)
+
+        if self.topk_before_score:
+            # TopK-then-Score: Select top-K by raw logits, then apply score_func (GPT-OSS style)
```
Comment (wwwjn) [inline torchtitan/models/moe/moe.py:329]: @shuhuayu  can we double check if this part is different between GPT-oss and Deepseek-v3?

Comment (shuhuayu) [reply torchtitan/models/moe/moe.py:329]: I'll look into it.

File: torchtitan/components/loss.py [hunk 6]
```diff
@@ -15,11 +15,18 @@
...
-    """Common cross-entropy loss function for Transformer models training."""
+    """Common cross-entropy loss function for Transformer models training.
+
+    Labels with CROSS_ENTROPY_IGNORE_INDEX (-100) are ignored, which is the
+    standard PyTorch convention for skipping padding tokens in loss computation.
+    """
     return torch.nn.functional.cross_entropy(
-        pred.flatten(0, 1).float(), labels.flatten(0, 1)
+        pred.flatten(0, 1).float(), labels.flatten(0, 1), ignore_index=CROSS_ENTROPY_IGNORE_INDEX
     )
```
Comment (fegin) [inline torchtitan/components/loss.py]: Why do you need this change? The default ignore_index is -100.

File: torchtitan/distributed/utils.py [hunk 7]
```diff
@@ -250,24 +250,18 @@ def maybe_enable_amp(
...
-                dtype=TORCH_DTYPE_MAP[mixed_precision_param],
-            )
+        # Enable autocast for all non-FSDP cases (DDP, TP-only, PP-only, single-device)
+        # torch.autocast works at the operator level and is orthogonal to parallelism strategy
+        logger.info("Mixed precision training is handled by AMP")
+        # pyrefly: ignore [bad-return]
+        return torch.autocast(
+            device_type,
+            dtype=TORCH_DTYPE_MAP[mixed_precision_param],
+        )
```
Comment (fegin) [inline torchtitan/distributed/utils.py]: iirc, AMP doesn't work well with PP or TP

**Review by fegin — CHANGES_REQUESTED** [review]
Thanks for contributing, but please split the PR into multiple ones. This PR contains both core components change and model changes, but many of them are unrelated to each other.

For checkpoint.py, DCP already has the ability to save and load the HF format. If the current implementation in DCP is insufficient, please file a draft PR or RFC or request to PyTorch. TorchTitan's philosophy is to reuse PyTorch core components as much as possible. DCP is one of them.

For optimizer.py, while I can see the need of multiple parameter groups, it is unclear that we need to create a different container. You can have multiple optimizers and multiple parameter groups. The reason why we need one model part to one optimizer is because of distributed state_dict design. We can probably eliminate the usage of distributed state_dict now, but that has not been done yet. The multiple parameter groups support along deserves a PR, seperating from CPUOffload optimzer.

For loss.py, the change is not needed. Or at I don't see why the change is required as it simply just explains the PyTorch core API usage.

For maybe_enable_amp, it also needs another PR because it changes the original logic.

Model changes deserver another PR(s) as well.

**Comment by eous** [comment]
> Thanks for your contribution! It would be better if you could help split this PR into several smaller pieces. It would be easier to review and discuss which feature to accept, for example:

Absolutely, I wasn't sure the best approach and after I saw gpt-oss got promoted out of experiments I figured I ran out of time trying to polish my improvements. I will submit these as seperate PR's. Additional comments inline below.

> 
>     * Optimizer CPU offload (How many users are actually interested in this comparing to other techniques to save memory?)

Unfortunately as a gpu-poor; I was having to get creative to train my persona series of models at interesting context lengths (https://huggingface.co/eousphoros/persona_zeta_20b_49k was trained using the code in this pr)

> 
>     * HuggingFacePackedDataset (Is this common practice in SFT?
> 

I can't say how common it is, as I am hobbyist but I have found it to have given me the best results for the datasets I am working with. (https://huggingface.co/datasets/eousphoros/persona_zeta_train dataset the model above was trained with)

>     * Differential learning rate / weight decay
> 

SFT fine-tuning optimizations

>     * GPT-OSS configs + MoE fix
> 

SFT fine-tuning correctness

>     * HuggingFace Checkpoint Loading (Why does this can not be handled properly with StateDictAdapter + DCP today?)

This is one of the older changes so I am not recalling all of the reasons for the changes atm but I do recall some issues with transposed vs not transposed alignment issues when loading huggingface mxfp4 and bf16 versions of gpt-oss into torch titan.

File: torchtitan/components/loss.py (hunk 6 above)
Comment (eous) [reply torchtitan/components/loss.py]: This was a mental note I added while I was working on the dataset loader changes, it can be backed out.

File: torchtitan/distributed/utils.py (hunk 7 above)
Comment (eous) [reply torchtitan/distributed/utils.py]: I haven't extensively tested this path (a full finetune) but it seemed to work with TP/EP but it can be backed out.

**Comment by eous** [comment]
This monolithic PR has been split into 7 focused PRs for easier review:

  https://github.com/pytorch/torchtitan/pull/2210 - Differential Learning Rate / Weight Decay

  Adds per-parameter-group learning rates and weight decay multipliers for fine-grained control during fine-tuning.

  Key changes:
  - LRMultipliers and WeightDecayMultipliers dataclasses in job_config.py
  - classify_parameters_for_groups() with regex-based pattern matching
  - OptimizersContainerWithParamGroups class
  - Pattern groups: bias, embeddings, output, attention, experts, routers, norms
  - Bias matched first, takes precedence for all *.bias parameters

  ---
  https://github.com/pytorch/torchtitan/pull/2211 - Attention Sink LSE Fix (Bug Fix)

  Fixes mathematically incorrect sigmoid-based attention sink implementation with proper LSE renormalization.

  Key changes:
  - Replace sigmoid(lse - sink) with exp(old_lse - new_lse) renormalization
  - Add clamping [-20, 0] for numerical stability
  - Mathematically equivalent to HuggingFace's concat+softmax approach

  ---
  https://github.com/pytorch/torchtitan/pull/2212 - MoE Routing Enhancements

  Adds GPT-OSS style MoE routing with TopK-then-Score strategy.

  Key changes:
  - topk_before_score: bool - GPT-OSS uses TopK-then-Score (vs DeepSeek's Score-then-TopK)
  - use_router_bias: bool - GPT-OSS has learned router biases
  - use_expert_bias: bool - GPT-OSS has learned expert biases
  - Fix histc → bincount for more reliable token counting

  ---
  https://github.com/pytorch/torchtitan/pull/2213 - AMP Fix for TP

  Removes overly conservative restriction that disabled mixed precision for TP-only configurations.

  Key changes:
  - Enable torch.autocast for TP-only training (operates at operator level, orthogonal to TP)
  - Clarify that PP uses schedule-based execution (unaffected)

  ---
  https://github.com/pytorch/torchtitan/pull/2214 - Expert Bias Support

  Adds optional expert biases (mlp1_bias, mlp2_bias) required for loading GPT-OSS pretrained models.

  Key changes:
  - use_expert_bias parameter in GptOssGroupedExperts
  - compute_dtype parameter for configurable compute precision
  - Cache invalidation after DTensor parallelization
  - Handle None biases in both for-loop and grouped_mm paths

  ---
  https://github.com/pytorch/torchtitan/pull/2215 - Router/Expert Bias Freezing

  Config options to freeze router and/or expert biases during MoE fine-tuning.

  Key changes:
  - freeze_router_bias: bool - preserve pretrained routing behavior
  - freeze_expert_bias: bool - preserve expert bias values
  - Warnings when freeze options enabled but no biases found

  ---
  https://github.com/pytorch/torchtitan/pull/2216 - YaRN RoPE Extensions

  Implements YaRN (Yet another RoPE extensioN) for context length extension (4096 → 131072 tokens).

  Key changes:
  - Frequency correction functions for smooth RoPE interpolation
  - precompute_rope_cache with YaRN scaling (rope_factor, beta_fast/slow)
  - mscale attention scaling factor (0.1 * ln(rope_factor) + 1.0) for numerical stability

  ---
  Not Included (per reviewer feedback)

  - CPU Offload Optimizer - needs separate discussion
  - HuggingFace Packed Dataset - questions if common practice
  - HuggingFace Checkpoint Loading - DCP should handle this
@wwwjn @fegin
